### ISSUE TITLE
Vec<T> implements PartialEq<[T; N]> but the other way round doesn't work

### ISSUE THREAD

**Issue by Ablesius** [rust-lang/rust#125243]
Hello,

it seems that there is an oversight (or possibly it's wanted behaviour, but it felt odd to me) in how vectors and arrays can be compared, based on `PartialEq` implementation.

I tried this code:

```rust
fn main() {
    let a = vec![1, 2, 3];
    let b = [1, 2, 3];
    
    assert_eq!(a, b);
    assert_eq!(b, a);
}
```
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3c95d140909d5ac62d769b8f663bcd47

I expected to see this happen: No errors because `assert_eq` should consider these two equal (or shouldn't it?)

Instead, this happened: 

```
   Compiling playground v0.0.1 (/playground)
error[E0277]: can't compare `[{integer}; 3]` with `Vec<{integer}>`
 --> src/main.rs:6:5
  |
6 |     assert_eq!(b, a);
  |     ^^^^^^^^^^^^^^^^ no implementation for `[{integer}; 3] == Vec<{integer}>`
  |
  = help: the trait `PartialEq<Vec<{integer}>>` is not implemented for `[{integer}; 3]`
  = note: this error originates in the macro `assert_eq` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider dereferencing here
 --> /playground/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/macros/mod.rs:40:35
  |
40|                 if !(*left_val == **right_val) {
  |                                   +

For more information about this error, try `rustc --explain E0277`.
error: could not compile `playground` (bin "playground") due to 1 previous error
```

### Meta
I also checked beta and nightly (on the playground, not on my own machine; please tell me if I should repeat it here), and the results are the same over all three of these.

`rustc --version --verbose`:
```
rustc 1.77.2 (25ef9e3d8 2024-04-09)
binary: rustc
commit-hash: 25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04
commit-date: 2024-04-09
host: x86_64-unknown-linux-gnu
release: 1.77.2
LLVM version: 17.0.6

```

<details><summary>Backtrace</summary>
<p>
(I did activate RUST_BACKTRACE=1 but the output was the same as before with the other runs)
</p>
</details>

**Comment by Ablesius** [comment]
Update:

```rust
let slice: &[i32] = &[1, 2, 3];
let vec = vec![1, 2, 3];

assert_eq!(slice, vec);
assert_eq!(vec, slice);
```

this works, which indicates further that it might just be an oversight that it doesn't work for comparing an array with a vec.

**Comment by RodBurman** [comment]
I have combined the code snippets above (along with some of my own work) to give:

    fn print_type<T>(name: &str, _: &T) { 
        println!("{name} is {:?}", std::any::type_name::<T>());
    }
    
    fn main() {
        let slice: &[i32] = &[1, 2, 3];
        let vec = vec![1, 2, 3];
        let array = [1, 2, 3];
    
        print_type("slice", &slice);
        print_type("vec", &vec);
        print_type("array", &array);
    
        assert_eq!(slice, vec);
        assert_eq!(vec, slice);
    
        assert_eq!(slice, array);
        assert_eq!(array, slice);
    
        assert_eq!(vec, array);
        // assert_eq!(array, vec);  // this line fails to compile error is E0277
    }

Using:
cargo 1.82.0 (8f40fc59f 2024-08-21)
release: 1.82.0
commit-hash: 8f40fc59fb0c8df91c97405785197f3c630304ea
commit-date: 2024-08-21
host: aarch64-apple-darwin
libgit2: 1.8.1 (sys:0.19.0 vendored)
libcurl: 8.7.1 (sys:0.4.74+curl-8.9.0 system ssl:(SecureTransport) LibreSSL/3.3.6)
ssl: OpenSSL 1.1.1w  11 Sep 2023
os: Mac OS 15.0.1 [64-bit]

It compiles and running it gives (to no ones surprise):
slice is "&[i32]"
vec is "alloc::vec::Vec<i32>"
array is "[i32; 3]"

But uncommenting assert_eq(array, vec) yields:

error[E0277]: can't compare `[i32; 3]` with `Vec<i32>`
  --> src/main.rs:21:5
   |
21 |     assert_eq!(array, vec);
   |     ^^^^^^^^^^^^^^^^^^^^^^ no implementation for `[i32; 3] == Vec<i32>`
   |
   = help: the trait `PartialEq<Vec<i32>>` is not implemented for `[i32; 3]`
   = note: this error originates in the macro `assert_eq` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider dereferencing here
  --> /Users/rod/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/macros/mod.rs:46:35
   |
46 |                 if !(*left_val == **right_val) {
   |  

Which is the original error. Having  X == Y but (effectively) Y != X does seem strange.
